### PR TITLE
add basic F# template

### PR DESCRIPTION
**Pull request by JaggerJo**
added basic F# Hello World template. Feedback is welcome

### CODE REVIEW

**Comment by kekekeks** [comment]
Why aren't you using XAML? It should work with F# out of the box. Even the upcoming XAML compiler should support F# just fine.

**Comment by JaggerJo** [comment]
I could add that but would prefer also providing a non XAML template.

What do you think about the 'template.json' file ? Have never built one before

**Comment by JaggerJo** [comment]
image

[link to the file](https://github.com/JaggerJo/avalonia-dotnet-templates/blob/master/app-fsharp-xaml/Main.fs)

You can reproduce in in my fork. Let me know if I can help with this

**Comment by grokys** [comment]
@JaggerJo I see you've added templates with XAML too now. In https://github.com/AvaloniaUI/Avalonia/issues/2281#issuecomment-468861420 you said:

> I had some problems with resource loading

What is the status of this now?

**Comment by JaggerJo** [comment]
@grokys I debugged it with Avalonia Sources and found the Issue. (Wrong Root Namespace)

Now everything works

File: app/App.xaml.cs [hunk 1]
```diff
@@ -7,6 +7,8 @@ public class App : Application
     {
         public override void Initialize()
         {
+            var x = this.GetType();
```
Comment (0xaryan) [inline app/App.xaml.cs]: `x` variables for the win.
I guess this committed by mistake?

File: app-fsharp-xaml/Program.fs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    inherit Application()
+    
+    override this.Initialize () =
+        let x = this.GetType()
+        AvaloniaXamlLoader.Load(this)
+ 
+module Program =
+
+    [<EntryPoint>]
+    let main argv =
```
Comment (kekekeks) [inline app-fsharp-xaml/Program.fs]: 1) this doesn't work with previewer, a separate BuildAvaloniaApp static method is needed
2) The preferred approach is a separate AppMain to ensure proper initialization order, see example here: https://github.com/AvaloniaUI/avalonia-dotnet-templates/blob/master/app/Program.cs

File: app-fsharp/Program.fs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+type App() =
+    inherit Application()
+    
+    override this.Initialize () =
+        this.Styles.Load "resm:Avalonia.Themes.Default.DefaultTheme.xaml?assembly=Avalonia.Themes.Default"
+        this.Styles.Load "resm:Avalonia.Themes.Default.Accents.BaseLight.xaml?assembly=Avalonia.Themes.Default"
+
+[<EntryPoint>]
+let main argv =
```
Comment (kekekeks) [inline app-fsharp/Program.fs]: Same here, separate `AppMain` and `BuildAvaloniaApp` are needed

File: app/App.xaml.cs (hunk 1 above)
Comment (JaggerJo) [reply app/App.xaml.cs]: yes, removed

File: app-fsharp-xaml/Program.fs (hunk 2 above)
Comment (JaggerJo) [reply app-fsharp-xaml/Program.fs]: Is resolved - previewer (VS 2019) works now

File: app-fsharp/Program.fs (hunk 3 above)
Comment (JaggerJo) [reply app-fsharp/Program.fs]: Completely deleted the non XAML template.

Maybe I will add a MVU (Model-View-Update) template in the future.

**Comment by JaggerJo** [comment]
@kekekeks @0xaryan @grokys 
Thank you for the feedback. I hope everything is resolved

image

**Comment by 0xaryan** [comment]
@JaggerJo Last changes LGTM and works as expected.
I've updated the readme and sent you a PR.
Thank you!

**Comment by 0xaryan** [comment]
I've added missing templates:

image

And a small PowerShell script which installs the templates:

image

It just needs @JaggerJo to review and merge my PR to his fork

**Comment by JaggerJo** [comment]
@0xaryan merged your PR.

**Comment by 0xaryan** [comment]
@JaggerJo Thank you

File: app-fsharp/AvaloniaAppTemplate.fsproj [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="MainWindow.fs" />
+    <Compile Include="App.fs" />
+    <Compile Include="Program.fs" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Compile Update="**\*.xaml.cs">
```
Comment (CameronAavik) [inline app-fsharp/AvaloniaAppTemplate.fsproj]: Just stumbled onto this PR as I was looking into doing F# with Avalonia and I believe this should be `*.xaml.fs`, not `cs`?

Comment (JaggerJo) [reply app-fsharp/AvaloniaAppTemplate.fsproj]: yes, it should be

**Comment by yatli** [comment]
wow, the code-behind generator handles F# now?

**Comment by JaggerJo** [comment]
@yatli avalonia does not rely on partial classes, there is (as far ad I know) nothing being generated.

**Comment by yatli** [comment]
@JaggerJo thanks for the background.
I see *.xaml.fs in the template, and I thought it was for the generated code.

So what’s the true purpose of it? :)

**Comment by yatli** [comment]
ok I see, a sample .xaml.fs is supplied with the template. Inside is the equivalent of a default C# codebehind class that does basic initialization stuff. The F# workflow will be fine without them.

**Comment by yatli** [comment]
To make the xaml files show up in the Visual Studio solution explorer, I have to modify the item group like this:

```fsproj
  <ItemGroup>
    <Compile Update="**\*.xaml.fs">
      <DependentUpon>%(Filename)</DependentUpon>
    </Compile>
    <AvaloniaResource Include="**\*.xaml">
      <SubType>Designer</SubType>
    </AvaloniaResource>
    <AvaloniaResource Include="Assets\*" />
    <Content Include="**\*.xaml">
      <SubType>Code</SubType>
    </Content>
  </ItemGroup>
```

**Comment by yatli** [comment]
hmm I guess the actual type does not matter -- as long as `AvaloniaResource` goes first.

**Comment by jkoritzinsky** [comment]
That documentation on the Page item type is specific to WPF. Avalonia only uses the AvaloniaResource item type for its XAML pages.

**Comment by 0xaryan** [comment]
@jkoritzinsky Correct, but using `AvaloniaResource` alone won't show Xaml files in the Visual Studio's Solution Explorer.

**Comment by jkoritzinsky** [comment]
Hmm looks like we need to add something to the VS extension. @grokys do you know/remember how to add a new item type in such a way that VS shows it?

**Comment by 0xaryan** [comment]
@jkoritzinsky If you mean either
```xml
<AvaloniaResource Include="**\*.xaml">
  <SubType>Designer</SubType>
  <Visible>true</Visible>
</AvaloniaResource>
```
Or
```xml
<AvaloniaResource Include="**\*.xaml">
  <SubType>Designer</SubType>
  <InProject>true</InProject>
</AvaloniaResource>
```
I think they don't work for F# projects. Maybe it's a bug, or I am doing something wrong.

**Comment by yatli** [comment]
IIRC for a vsix, the type provider must be registered against the project GUID type.
For F# it is `{F2A71F9B-5D33-465A-A702-920D77279786}`
FYI https://www.codeproject.com/Reference/720512/List-of-Visual-Studio-Project-Type-GUIDs

**Comment by jmacato** [comment]
@JaggerJo What's our gameplan on this F# templates? Would be great if this gets merged soon or before 0.9 final

**Comment by JaggerJo** [comment]
@jmacato Yep - that would be good. 

What's preventing this from merging ? The missing extension support ?

Sorry, lost track on this one.

**Comment by FoggyFinder** [comment]
What's the status of the PR?

**Comment by grokys** [comment]
The repository containing this branch seems to have disappeared meaning that I can't check it out locally. @JaggerJo any idea what happened there?

In addition, we've merged F# templates into the VS extension (https://github.com/AvaloniaUI/AvaloniaVS/pull/175) -- which templates are more likely to be up-to-date? Shall I port the ones from the VS extension over?

**Comment by JaggerJo** [comment]
> The repository containing this branch seems to have disappeared meaning that I can't check it out locally. @JaggerJo any idea what happened there?

Good question, I don't know TBH.

> In addition, we've merged F# templates into the VS extension ([AvaloniaUI/AvaloniaVS#175](https://github.com/AvaloniaUI/AvaloniaVS/pull/175)) -- which templates are more likely to be up-to-date? Shall I port the ones from the VS extension over?

Might be good so the templates are the same.

**Comment by grokys** [comment]
Ok, i'll close this and add the F# templates in a new PR.

**Pull request closed without merging**